test/uri_util_test.rb

```diff
@@ -0,0 +1,75 @@
+require File.expand_path('../support/test_helper', __FILE__)
+
+class UriUtilTest < Minitest::Test
+  def test_ref_fragment_path
+    uri = '#some-thing'
+    base = 'http://www.example.com/foo/#bar'
+
+    assert_equal Addressable::URI.parse('http://www.example.com/foo/#some-thing'), JSON::Util::URI.normalize_ref(uri, base)
+    assert_equal Addressable::URI.parse('http://www.example.com/foo/#'), JSON::Util::URI.absolutize_ref(uri, base)
+  end
+
+  def test_ref_file_path
+    uri = '/some/thing'
+    base = 'http://www.example.com/foo/#bar'
+
+    assert_equal Addressable::URI.parse('http://www.example.com/some/thing#'), JSON::Util::URI.normalize_ref(uri, base)
+    assert_equal Addressable::URI.parse('http://www.example.com/some/thing#'), JSON::Util::URI.absolutize_ref(uri, base)
+  end
+
+  def test_ref_uri
+    uri = 'http://foo-bar.com'
+    base = 'http://www.example.com/foo/#bar'
+
+    assert_equal Addressable::URI.parse('http://foo-bar.com/#'), JSON::Util::URI.normalize_ref(uri, base)
+    assert_equal Addressable::URI.parse('http://foo-bar.com/#'), JSON::Util::URI.absolutize_ref(uri, base)
+  end
+
+  def test_ref_uri_with_path
+    uri = 'http://foo-bar.com/some/thing'
+    base = 'http://www.example.com/foo/#bar'
+
+    assert_equal Addressable::URI.parse('http://foo-bar.com/some/thing#'), JSON::Util::URI.normalize_ref(uri, base)
+    assert_equal Addressable::URI.parse('http://foo-bar.com/some/thing#'), JSON::Util::URI.absolutize_ref(uri, base)
+  end
+
+  def test_ref_uri_with_fragment
+    uri = 'http://foo-bar.com/some/thing#foo'
+    base = 'http://www.example.com/foo/#bar'
+
+    assert_equal Addressable::URI.parse('http://foo-bar.com/some/thing#foo'), JSON::Util::URI.normalize_ref(uri, base)
+    assert_equal Addressable::URI.parse('http://foo-bar.com/some/thing#'), JSON::Util::URI.absolutize_ref(uri, base)
+  end
+
+  def test_ref_uri_with_fragment_and_base_with_no_fragment
+    uri = 'http://foo-bar.com/some/thing#foo'
+    base = 'http://www.example.com/foo'
+
+    assert_equal Addressable::URI.parse('http://foo-bar.com/some/thing#foo'), JSON::Util::URI.normalize_ref(uri, base)
+    assert_equal Addressable::URI.parse('http://foo-bar.com/some/thing#'), JSON::Util::URI.absolutize_ref(uri, base)
+  end
+
+  def test_ref_relative_path
+    uri = 'hello/world'
+    base = 'http://www.example.com/foo/#bar'
+
+    assert_equal Addressable::URI.parse('http://www.example.com/foo/hello/world#'), JSON::Util::URI.normalize_ref(uri, base)
+    assert_equal Addressable::URI.parse('http://www.example.com/foo/hello/world#'), JSON::Util::URI.absolutize_ref(uri, base)
+  end
+
+  def test_ref_addressable_uri_with_host
+    uri = Addressable::URI.new(:host => 'foo-bar.com')
+    base = 'http://www.example.com/foo/#bar'
+
+    assert_equal Addressable::URI.parse('http://www.example.com/foo-bar.com#'), JSON::Util::URI.normalize_ref(uri, base)
```

Comment:
I've just tried to document the existing behaviour for now. I've added some extra tests that do include a schema and this changes the behaviour entirely!